src/gradient_descent.c

```diff
@@ -0,0 +1,380 @@
+/// the initialPoint along the given direction
+/// @param func Initial function to transform
+/// @param x Value at which to evaluate the function
+/// @param initialPoint Initial point to base the movement of
+/// @param direction Direction in which to move by x
+/// @param n Number of dimensions of the original function
+/// @return Function value after moving by x from the initial point in the given
+/// direction
+static gradient_real oneDimension(function func, gradient_real x,
+                                  gradient_real initialPoint[],
+                                  gradient_real direction[], int n) {
+  gradient_real *point = malloc(n * sizeof(gradient_real));
+
+  // Compute new point when moving by x in the given direction
+  for (int i = 0; i < n; i++) {
+    point[i] = initialPoint[i] + x * direction[i];
+  }
+
+  // Compute function value at new point
+  gradient_real value = (*func)(point);
+
+  free(point);
+
+  return value;
+}
+
+/// @brief Determines the function minimum from 3 points initially bracketing
+/// the minimum. Uses a mix of parabolic interpolation and Golden Section
+/// search.
+/// @param xMin Abscissa of the minimum. Will contain the Abscissa of the
+/// minimum after function executes
+/// @param bracket Initial 3 points bracketing the minimum
+/// @param func Function transforming an N dimension function to a 1 dimension
+/// function
+/// @param f Function we want to transfor to a 1 dimension function
+/// @param initialPoint Initial point to base the movement of
+/// @param direction Direction in which to move
+/// @param n Number of dimensions of the original function
+/// @param tol Tolerance for the minimum estimate. Should be no smaller then the
+/// square-root of the machine floating point precision
+/// @return Value of the original function at the minimum found
+static gradient_real brent(gradient_real *xMin, Bracket bracket,
+                           f1dimension func, function f,
+                           gradient_real initialPoint[],
+                           gradient_real direction[], int n,
+                           gradient_real tol) {
+
+  // Distance moved
+  gradient_real distance = 0.0;
+
+  // Distance moved previously
+  gradient_real previousDistance = 0.0;
+
+  // Minimum is bracketed between a and b
+  gradient_real a = (bracket.a < bracket.c ? bracket.a : bracket.c);
+  gradient_real b = (bracket.a > bracket.c ? bracket.a : bracket.c);
+
+  // xMin : Point with the smallest function value found so far
+  // secondXMin : Point with the second smallest function value found so far
+  // prevSecondXMin : Previous value of secondXMin
+  gradient_real prevSecondXMin, secondXMin;
+  *xMin = prevSecondXMin = secondXMin = bracket.b;
+
+  // Value of the function at the above points
+  gradient_real fXMin, fPrevSecondXMin, fSecondXMin;
+  fXMin = fPrevSecondXMin = fSecondXMin =
+      (*func)(f, *xMin, initialPoint, direction, n);
```

Comment:
There exists an easier way to use your function
`fxmin = fPrevSecondXMin = fSecondXMin = func(f, *xMin, initialPoint, direction, n);`